packages/meditrak-server/src/routes/mapOverlayGroups/assertMapOverlayGroupsPermissions.js

```diff
@@ -0,0 +1,213 @@
+      const result = await hasMapOverlayGetPermissions(
+        accessPolicy,
+        models,
+        mapOverlayRelation.child_id,
+      );
+      if (result.result) {
+        return result;
+      }
+    }
+  }
+
+  if (mapOverlayGroupRelations.length) {
+    for (const mapOverlayRelation of mapOverlayRelations) {
+      const result = await hasMapOverlayGroupGetPermissions(
+        accessPolicy,
+        models,
+        mapOverlayRelation.child_id,
+      );
+      if (result.result) {
+        return result;
+      }
+    }
+  }
+
+  return {
+    result: false,
+    errorMessage: `Requires access to all the child items of map overlay group with id "${mapOverlayGroupId}"`,
+  };
+};
+
+export const hasMapOverlayGroupEditPermissions = async (
+  accessPolicy,
+  models,
+  mapOverlayGroupId,
+) => {
+  const mapOverlayGroup = await models.mapOverlayGroup.findById(mapOverlayGroupId);
+  if (!mapOverlayGroup) {
+    return {
+      result: false,
+      errorMessage: `No map overlay group exists with id "${mapOverlayGroupId}"`,
+    };
+  }
+
+  const childMapOverlayRelations = await models.mapOverlayGroupRelation.find({
+    map_overlay_group_id: mapOverlayGroupId,
+  });
+
+  if (!childMapOverlayRelations.length) {
+    return true;
+  }
+
+  const mapOverlayRelations = childMapOverlayRelations.filter(
+    r => r.child_type === models.mapOverlayGroupRelation.RelationChildTypes.MAP_OVERLAY,
+  );
+  const mapOverlayGroupRelations = childMapOverlayRelations.filter(
+    r => r.child_type === models.mapOverlayGroupRelation.RelationChildTypes.MAP_OVERLAY_GROUP,
+  );
+
+  if (mapOverlayRelations.length) {
+    for (const mapOverlayRelation of mapOverlayRelations) {
+      const result = await hasMapOverlayEditPermissions(
+        accessPolicy,
+        models,
+        mapOverlayRelation.child_id,
+      );
+      if (!result.result) {
+        return result;
+      }
+    }
+  }
+
+  if (mapOverlayGroupRelations.length) {
+    for (const mapOverlayRelation of mapOverlayRelations) {
+      const result = await hasMapOverlayGroupEditPermissions(
+        accessPolicy,
+        models,
+        mapOverlayRelation.child_id,
+      );
+      if (!result.result) {
+        return result;
+      }
+    }
+  }
+
+  return {
+    result: true,
+  };
+};
+
+export const assertMapOverlayGroupsGetPermissions = async (
+  accessPolicy,
+  models,
+  mapOverlayGroupId,
+) => {
+  const result = await hasMapOverlayGroupGetPermissions(accessPolicy, models, mapOverlayGroupId);
+  if (result.result) {
+    return true;
+  }
+
+  throw new Error(result.errorMessage);
+};
+
+export const assertMapOverlayGroupsEditPermissions = async (accessPolicy, models, mapOverlayId) => {
+  const result = await hasMapOverlayGroupEditPermissions(accessPolicy, models, mapOverlayId);
+
+  if (result.result) {
+    return true;
+  }
+
+  throw new Error(result.errorMessage);
+};
+
+const findPermittedMapOverlayGroupIds = (
```

Comment:
Nice recursive function!! 👏

Recursion is notoriously complicated, so a comment that explains the general recursion logic would be beneficial. What would be especially helpful is describing how we traverse the relation tree (if I got this right, bottom-up until we reach the root nodes)

Reply:
I should have done that 😿